examples/file-sharing.rs

```diff
@@ -0,0 +1,639 @@
+        name: String,
+    },
+    Get {
+        #[structopt(long)]
+        name: String,
+    },
+}
+
+/// The network module, encapsulating all network related logic.
+mod network {
+    use super::*;
+    use async_trait::async_trait;
+    use futures::channel::{mpsc, oneshot};
+    use futures::future::BoxFuture;
+    use futures::stream::BoxStream;
+    use libp2p::core::upgrade::{read_length_prefixed, write_length_prefixed, ProtocolName};
+    use libp2p::identity;
+    use libp2p::identity::ed25519;
+    use libp2p::kad::record::store::MemoryStore;
+    use libp2p::kad::{GetProvidersOk, Kademlia, KademliaEvent, QueryResult};
+    use libp2p::multiaddr::Protocol;
+    use libp2p::request_response::{
+        ProtocolSupport, RequestResponse, RequestResponseCodec, RequestResponseEvent,
+        RequestResponseMessage, ResponseChannel,
+    };
+    use libp2p::swarm::{SwarmBuilder, SwarmEvent};
+    use libp2p::{NetworkBehaviour, Swarm};
+    use std::collections::{HashMap, HashSet};
+    use std::iter;
+
+    /// Creates the network components, namely:
+    ///
+    /// - The network client to interact with the network layer from anywhere
+    ///   within your application.
+    ///
+    /// - The network event stream, e.g. for incoming requests.
+    ///
+    /// - The network task driving the network itself.
+    pub fn new(
+        listen_address: Option<Multiaddr>,
+        secret_key_seed: Option<u8>,
+    ) -> Result<(Client, BoxStream<'static, Event>, BoxFuture<'static, ()>), Box<dyn Error>> {
+        // Create a public/private key pair, either random or based on a seed.
+        let id_keys = match secret_key_seed {
+            Some(seed) => {
+                let mut bytes = [0u8; 32];
+                bytes[0] = seed;
+                let secret_key = ed25519::SecretKey::from_bytes(&mut bytes).expect(
+                    "this returns `Err` only if the length is wrong; the length is correct; qed",
+                );
+                identity::Keypair::Ed25519(secret_key.into())
+            }
+            None => identity::Keypair::generate_ed25519(),
+        };
+        let peer_id = id_keys.public().to_peer_id();
+
+        // Build the Swarm, connecting the lower layer transport logic with the
+        // higher layer network behaviour logic.
+        let mut swarm = SwarmBuilder::new(
+            futures::executor::block_on(libp2p::development_transport(id_keys))?,
+            ComposedBehaviour {
+                kademlia: Kademlia::new(peer_id, MemoryStore::new(peer_id)),
+                request_response: RequestResponse::new(
+                    FileExchangeCodec(),
+                    iter::once((FileExchangeProtocol(), ProtocolSupport::Full)),
+                    Default::default(),
+                ),
+            },
+            peer_id,
+        )
+        .build();
+
+        // In case a listen address was provided use it, otherwise listen on any
+        // address.
+        match listen_address {
```

Comment:
I'd do this via an new method `Client::start_listening` & `Command::Start_Listening`, to be consistent.